File: app/src/main/java/com/onemoresecret/msg_fragment_plugins/MsgPluginWiFiPairing.java
package com.onemoresecret.msg_fragment_plugins;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.wifi.WifiInfo;
import android.util.Log;

import androidx.fragment.app.Fragment;

import com.onemoresecret.MainActivity;
import com.onemoresecret.MessageFragment;
import com.onemoresecret.OmsDataInputStream;
import com.onemoresecret.OutputFragment;
import com.onemoresecret.R;
import com.onemoresecret.WiFiPairingFragment;
import com.onemoresecret.crypto.Base58;
import com.onemoresecret.crypto.RSAUtil;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.net.Inet4Address;
import java.net.ServerSocket;
import java.nio.ByteBuffer;
import java.util.Arrays;

public class MsgPluginWiFiPairing extends MessageFragmentPlugin {
    private static final String TAG = MsgPluginWiFiPairing.class.getSimpleName();
    private static final long ttl_default = 12L * 3600_000L; //12 hours

    public record IpAndPort(byte[] ipAddress, int port, byte[] responseCode) {
    }

    public MsgPluginWiFiPairing(MessageFragment messageFragment, byte[] messageData) throws Exception {
        super(messageFragment);
        try (ByteArrayInputStream bais = new ByteArrayInputStream(messageData);
             var dataInputStream = new OmsDataInputStream(bais)) {

            //(1) Application ID - already read, not present here

            //(2) Request ID
            var requestId = dataInputStream.readString();

            //(3) RSA public key - PC
            var rsaPublicKeyMaterial = dataInputStream.readByteArray();

            //(4) RSA privateKey to protect communication
            var rsaPrivateKeyMaterial = dataInputStream.readByteArray();

            //get IP address
            var ipAndPort = getIpAndPort(context);

            context.getMainExecutor().execute(() -> {
                var responseCodeBase58 = Base58.encode(ipAndPort.responseCode);
                ((WiFiPairingFragment) getMessageView()).setData(
                        requestId,
                        responseCodeBase58,
                        () -> {
                            ((OutputFragment) getOutputView()).setMessage(responseCodeBase58 + "\n", "Response Code");
                            ((MainActivity) context).setWiFiComm(
                                    new MainActivity.WiFiComm(
                                            ipAndPort.ipAddress,
                                            ipAndPort.port,
                                            rsaPrivateKeyMaterial,
                                            System.currentTimeMillis() + ttl_default),
                                    rsaPrivateKeyMaterial);
                        });
            });
        }
    }

    public static IpAndPort getIpAndPort(Context context) throws IOException {
        var connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);

        var network = connectivityManager.getActiveNetwork();
        if (network == null) {
            throw new IOException(context.getString(R.string.no_network));
        }
        var networkCapabilities = connectivityManager.getNetworkCapabilities(network);
        if (networkCapabilities == null)
            throw new IOException(context.getString(R.string.not_on_wifi_network));
        var transportInfo = networkCapabilities.getTransportInfo();
        if (!(transportInfo instanceof WifiInfo)) {
            throw new IOException(context.getString(R.string.not_on_wifi_network));
        }
        var linkProperties = connectivityManager.getLinkProperties(network);
        if (linkProperties != null) {
            for (var linkAddress : linkProperties.getLinkAddresses()) {
                var inetAddress = linkAddress.getAddress();
                if (!(inetAddress instanceof Inet4Address)) continue;

                var ipAddress = inetAddress.getAddress();
                Log.d(TAG, "IP: " + inetAddress.getHostAddress() + " = " + Arrays.toString(ipAddress));

                try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
                     ServerSocket serverSocket = new ServerSocket(0)) {

                    var port = serverSocket.getLocalPort();

                    //(1) IP address as integer value
                    baos.write(ipAddress);

                    //(2) port
                    var iArr = ByteBuffer.allocate(4).putInt(port).array();

                    //copy only lower portion, ports range 0...65535
                    baos.write(Arrays.copyOfRange(iArr, 2, iArr.length));

                    var bArr = baos.toByteArray();

                    return new IpAndPort(ipAddress, port, bArr);
                }
            }
        }

        throw new IOException("Unsupported network (not IP4)");
    }

    @Override
    public Fragment getMessageView() {
        if (messageView == null) messageView = new WiFiPairingFragment();
        return messageView;
    }

    @Override
    public void showBiometricPromptForDecryption() {
        //not used
    }
}

File: app/src/main/java/com/onemoresecret/msg_fragment_plugins/MsgPluginWiFiPairing.kt
package com.onemoresecret.msg_fragment_plugins

import android.content.Context
import android.net.ConnectivityManager
import android.net.wifi.WifiInfo
import android.util.Log
import androidx.compose.animation.AnimatedContent
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.fragment.app.FragmentActivity
import com.onemoresecret.MainActivity
import com.onemoresecret.OmsDataInputStream
import com.onemoresecret.R
import com.onemoresecret.composable.OutputScreen
import com.onemoresecret.composable.OutputViewModel
import com.onemoresecret.crypto.Base58
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.net.Inet4Address
import java.net.ServerSocket
import java.nio.ByteBuffer
import java.util.Arrays
import kotlinx.coroutines.flow.MutableStateFlow


class MsgPluginWiFiPairing(
    activity: FragmentActivity,
    messageData: ByteArray,
    hiddenState: MutableStateFlow<Boolean>,
    onNavigateBack: () -> Unit
) : MessageFragmentPlugin(activity, hiddenState, onNavigateBack) {

    private var requestId by mutableStateOf("")
    private var responseCode by mutableStateOf("")
    private var onStartClick: (() -> Unit)? by mutableStateOf(null)

    private val outputViewModel = OutputViewModel(preferences)

    init {
        OmsDataInputStream(ByteArrayInputStream(messageData)).use { dataInputStream ->
            requestId = dataInputStream.readString()
            val rsaPublicKeyMaterial = dataInputStream.readByteArray()
            val rsaPrivateKeyMaterial = dataInputStream.readByteArray()

            val ipAndPort = getIpAndPort(context)

            context.mainExecutor.execute {
                responseCode = Base58.encode(ipAndPort.responseCode)
                onStartClick = {
                    outputViewModel.setMessage(responseCode, "Response Code")
                    (context as MainActivity).setWiFiComm(
                        MainActivity.WiFiComm(
                            ipAndPort.ipAddress,
                            ipAndPort.port,
                            rsaPublicKeyMaterial,
                            System.currentTimeMillis() + TTL_DEFAULT
                        ),
                        rsaPrivateKeyMaterial
                    )
                }
            }
        }
    }

    @Composable
    override fun MessageView(hiddenState: Boolean) {
        WiFiPairingScreen(
            requestId = requestId,
            responseCode = if (hiddenState) "●".repeat(responseCode.length) else responseCode,
            onStartClick = onStartClick
        )
    }

    @Composable
    override fun TopBarActions() {

    }

    @Composable
    override fun OutputView() {
        OutputScreen(outputViewModel = outputViewModel)
    }

    override fun showBiometricPromptForDecryption() {
    }

    data class IpAndPort(val ipAddress: ByteArray, val port: Int, val responseCode: ByteArray)

    companion object {
        private const val TTL_DEFAULT = 12L * 3600_000L //12 hours
        private val TAG = MsgPluginWiFiPairing::class.java.simpleName

        @Throws(IOException::class)
        fun getIpAndPort(context: Context): IpAndPort {
            val connectivityManager =
                context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

            val network = connectivityManager.activeNetwork
                ?: throw IOException(context.getString(R.string.no_network))
            val networkCapabilities = connectivityManager.getNetworkCapabilities(network)
                ?: throw IOException(context.getString(R.string.not_on_wifi_network))
            val transportInfo = networkCapabilities.transportInfo
            if (transportInfo !is WifiInfo) {
                throw IOException(context.getString(R.string.not_on_wifi_network))
            }

            val linkProperties = connectivityManager.getLinkProperties(network)
            if (linkProperties != null) {
                for (linkAddress in linkProperties.linkAddresses) {
                    val inetAddress = linkAddress.address
                    if (inetAddress !is Inet4Address) continue

                    val ipAddress = inetAddress.address
                    Log.d(TAG, "IP: " + inetAddress.hostAddress + " = " + Arrays.toString(ipAddress))

                    ByteArrayOutputStream().use { baos ->
                        ServerSocket(0).use { serverSocket ->
                            val port = serverSocket.localPort
                            baos.write(ipAddress)
                            val iArr = ByteBuffer.allocate(4).putInt(port).array()
                            baos.write(Arrays.copyOfRange(iArr, 2, iArr.size))
                            val bArr = baos.toByteArray()
                            return IpAndPort(ipAddress, port, bArr)
                        }
                    }
                }
            }

            throw IOException("Unsupported network (not IP4)")
        }
    }
}



@Composable
fun WiFiPairingScreen(
    requestId: String,
    responseCode: String,
    onStartClick: (() -> Unit)?
) {
    var isPairingAccepted by remember { mutableStateOf(false) }

    WiFiPairing(
        requestId = requestId,
        responseCode = responseCode,
        isPairingAccepted = isPairingAccepted,
        onConfirm = {
            isPairingAccepted = true
            onStartClick?.invoke()
        }
    )
}

@Composable
fun WiFiPairing(
    requestId: String,
    responseCode: String,
    isPairingAccepted: Boolean,
    onConfirm: () -> Unit
) {
    AnimatedContent(targetState = isPairingAccepted, label = "pairingStateAnimation") { accepted ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Top
        ) {
            if (accepted) {
                Text(
                    text = stringResource(id = R.string.pairing_info_response),
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(top = 8.dp)
                )

                Text(
                    text = responseCode,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(top = 8.dp),
                    style = MaterialTheme.typography.headlineMedium,
                    textAlign = TextAlign.Center
                )
            } else {
                Text(
                    text = stringResource(id = R.string.pairing_info_request),
                    modifier = Modifier.fillMaxWidth()
                )

                Text(
                    text = requestId,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(top = 8.dp),
                    style = MaterialTheme.typography.headlineMedium,
                    textAlign = TextAlign.Center
                )

                Button(
                    onClick = onConfirm,
                    modifier = Modifier.padding(top = 8.dp)
                ) {
                    Text(
                        text = stringResource(
                            id = R.string.accept_pairing
                        )
                    )
                }
            }
        }
    }
}
